phoenix-core/src/it/java/org/apache/phoenix/end2end/RowTimestampEpochFunctionIT.java

```diff
@@ -0,0 +1,161 @@
+                assertEquals(rs.getDate(1).getTime(), timeStamp);
+                ++rowCount;
+            }
+            assertEquals(expectedRowCount, rowCount);
+        }
+    }
+
+    @Test
+    public void testRowTimestampDefault() throws Exception {
+
+        String tableName =  generateUniqueName();
+        try (Connection conn = DriverManager.getConnection(getUrl())) {
+            String ddl = "CREATE TABLE IF NOT EXISTS " + tableName
+                    + " (PK INTEGER NOT NULL PRIMARY KEY, KV1 VARCHAR, KV2 VARCHAR)" + this.tableDDLOptions;
+            conn.createStatement().execute(ddl);
+
+            String dml = "UPSERT INTO " + tableName + " (PK, KV1, KV2) VALUES (?, ?, ?)";
+            PreparedStatement stmt = conn.prepareStatement(dml);
+
+            int count = 5;
+            for (int id = 0; id < count; ++id) {
+                stmt.setInt(1, id);
+                stmt.setString(2, "KV1_" + id);
+                stmt.setString(3, "KV2_" + id);
+                stmt.executeUpdate();
+            }
+            conn.commit();
+
+            String dql = "SELECT ROW_TIMESTAMP_EPOCH() FROM " + tableName;
+
+            ResultSet rs = conn.createStatement().executeQuery(dql);
+            // verify row timestamp returned by the query matches the empty column cell timestamp
+            verifyHbaseRowTimestamp(tableName, rs, count);
+
+            // update one row
+            conn.createStatement().execute("UPSERT INTO " + tableName
+                    + " (PK, KV1) VALUES (2, 'KV1_foo')");
+            conn.commit();
+
+            rs = conn.createStatement().executeQuery(dql);
+            // verify again after update
+            verifyHbaseRowTimestamp(tableName, rs, count);
+        }
+    }
+
+    @Test
+    public void testRowTimestampColumn() throws Exception {
+        String tableName =  generateUniqueName();
+        try (Connection conn = DriverManager.getConnection(getUrl())) {
+            String ddl = "CREATE TABLE IF NOT EXISTS " + tableName
+                    + " (PK1 INTEGER NOT NULL, PK2 DATE NOT NULL, KV1 VARCHAR, KV2 VARCHAR"
+                    + " CONSTRAINT PK PRIMARY KEY(PK1, PK2 ROW_TIMESTAMP))" + this.tableDDLOptions;
+            conn.createStatement().execute(ddl);
+
+            String dml = "UPSERT INTO " + tableName + " (PK1, PK2, KV1, KV2) VALUES (?, ?, ?, ?)";
+
+            long rowTimestamp = EnvironmentEdgeManager.currentTimeMillis();
+            Date rowTimestampDate = new Date(rowTimestamp);
+            PreparedStatement stmt = conn.prepareStatement(dml);
+
+            int count = 5;
+            for (int id = 0; id < count; ++id) {
+                stmt.setInt(1, id);
+                stmt.setDate(2, rowTimestampDate);
+                stmt.setString(3, "KV1_" + id);
+                stmt.setString(4, "KV2_" + id);
+                stmt.executeUpdate();
+            }
+            conn.commit();
+
+            String dql = "SELECT ROW_TIMESTAMP_EPOCH() FROM " + tableName;
```

Comment:
could you please add one test for using this function in the WHERE clause?

Reply:
sure